File: server/agent/ingestionService.ts
import * as fs from 'fs';
import * as path from 'path';
import * as dotenv from 'dotenv';
import { GoogleGenAI } from '@google/genai';
import { dbQuery, dbPool } from '../db/client';
import { extractTextFromFile } from './documentParser';

dotenv.config();

const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
if (!GEMINI_API_KEY) {
  console.error('Error: La variable de entorno GEMINI_API_KEY no está configurada.');
  process.exit(1);
}

// Inicializar el SDK de Google Gen AI
const ai = new GoogleGenAI({ apiKey: GEMINI_API_KEY });

/**
 * Divide un texto plano en fragmentos (chunks) más pequeños con solapamiento (overlap)
 */
function chunkText(text: string, chunkSize: number = 1000, overlap: number = 150): string[] {
  const chunks: string[] = [];
  let currentIndex = 0;
  
  while (currentIndex < text.length) {
    let endIndex = currentIndex + chunkSize;
    
    // Si no estamos al final del texto, intentar cortar en un espacio en blanco o salto de línea
    if (endIndex < text.length) {
      const lastSpace = text.lastIndexOf(' ', endIndex);
      if (lastSpace > currentIndex) {
        endIndex = lastSpace;
      }
    }
    
    const chunk = text.substring(currentIndex, endIndex).trim();
    if (chunk.length > 0) {
      chunks.push(chunk);
    }
    
    currentIndex = endIndex - overlap;
    if (currentIndex >= text.length || chunkSize <= overlap) {
      break;
    }
  }
  
  return chunks;
}

/**
 * Procesa un archivo CSV y genera fragmentos agrupados por registros (filas) para preservar su estructura tabular
 */
function chunkCsvText(csvText: string, rowsPerChunk: number = 5): string[] {
  const lines = csvText.split('\n').filter(line => line.trim().length > 0);
  if (lines.length === 0) return [];
  
  const header = lines[0]; // Extraer el encabezado del CSV
  const dataLines = lines.slice(1);
  const chunks: string[] = [];
  
  for (let i = 0; i < dataLines.length; i += rowsPerChunk) {
    const chunkLines = dataLines.slice(i, i + rowsPerChunk);
    // Agrupar filas incluyendo el encabezado al inicio de cada chunk para dar contexto al modelo
    const chunk = [header, ...chunkLines].join('\n');
    chunks.push(chunk);
  }
  
  return chunks;
}

/**
 * Determina la categoría del documento según su nombre de archivo
 */
function getDocumentCategory(fileName: string): string {
  const lowerName = fileName.toLowerCase();
  if (lowerName.includes('entorno') || lowerName.includes('configuracion')) {
    return 'Entorno';
  }
  if (lowerName.includes('estilo') || lowerName.includes('practicas')) {
    return 'Estilo';
  }
  if (lowerName.includes('modulo') || lowerName.includes('arquitectura')) {
    return 'Arquitectura';
  }
  return 'General';
}

/**
 * Genera el embedding de un fragmento de texto usando el modelo de embeddings de Gemini
 */
async function generateEmbedding(text: string): Promise<number[]> {
  try {
    const response = await ai.models.embedContent({
      model: 'text-embedding-004',
      contents: text,
    });
    
    if (response.embedding?.values) {
      return response.embedding.values;
    }
    throw new Error('No se recibieron valores de embedding desde la API de Gemini.');
  } catch (error) {
    console.error('Error al llamar a la API de embeddings de Gemini:', error);
    throw error;
  }
}

/**
 * Pipeline principal de ingesta de documentos locales
 */
async function startIngestion() {
  console.log('Iniciando pipeline de ingesta de documentos RAG...');
  
  const kbDir = path.join(__dirname, '..', '..', 'knowledge-base');
  
  if (!fs.existsSync(kbDir)) {
    console.error(`El directorio de base de conocimientos no existe: ${kbDir}`);
    process.exit(1);
  }
  
  const files = fs.readdirSync(kbDir);
  console.log(`Archivos encontrados en la base de conocimientos: ${files.join(', ')}`);
  
  // Limpiar chunks anteriores para evitar duplicados en la ingesta inicial
  try {
    console.log('Limpiando base de datos de fragmentos anteriores...');
    await dbQuery('DELETE FROM document_chunks');
  } catch (err) {
    console.error('No se pudo limpiar la base de datos (asegúrate de correr primero las migraciones):', err);
    process.exit(1);
  }

  for (const file of files) {
    const filePath = path.join(kbDir, file);
    const category = getDocumentCategory(file);
    
    console.log(`\nProcesando archivo: "${file}" [Categoría: ${category}]`);
    
    try {
      // 1. Extraer texto según formato
      const fullText = await extractTextFromFile(filePath);
      
      // 2. Fragmentar texto según el formato (CSV recibe un chunking diferente)
      let chunks: string[] = [];
      const ext = path.extname(file).toLowerCase();
      if (ext === '.csv') {
        chunks = chunkCsvText(fullText, 4); // Agrupar de 4 en 4 filas con encabezado
      } else {
        chunks = chunkText(fullText, 800, 120); // Trozos estándar para Markdown/texto
      }
      
      console.log(`Archivo fragmentado en ${chunks.length} partes.`);
      
      // 3. Generar embeddings y guardar en base de datos
      for (let i = 0; i < chunks.length; i++) {
        const chunkContent = chunks[i];
        console.log(`Generando embedding e insertando fragmento ${i + 1}/${chunks.length}...`);
        
        // Obtener el vector embedding
        const vector = await generateEmbedding(chunkContent);
        
        // Convertir el array numérico de JS a formato vector de PostgreSQL: '[0.123, -0.456, ...]'
        const dbVectorFormat = `[${vector.join(',')}]`;
        
        // Insertar en la base de datos
        await dbQuery(
          `INSERT INTO document_chunks (documento_nombre, categoria, contenido, embedding)
           VALUES ($1, $2, $3, $4)`,
          [file, category, chunkContent, dbVectorFormat]
        );
      }
      
      console.log(`¡Archivo "${file}" indexado exitosamente!`);
    } catch (err) {
      console.error(`Error crítico procesando el archivo "${file}":`, err);
    }
  }
  
  console.log('\n=========================================');
  console.log('¡Pipeline de ingesta RAG finalizado con éxito!');
  console.log('Todos los documentos se han procesado y guardado.');
  console.log('=========================================');
  
  // Finalizar pool de base de datos
  await dbPool.end();
}

// Ejecutar si se llama directamente desde node/ts-node
if (require.main === module) {
  startIngestion();
}

File: agente-alura/agent/ingestion_service.py
import os
import sys
import datetime
import json

# Evitar errores de codificación en consolas Windows (CP1252)
if sys.platform.startswith('win'):
    try:
        sys.stdout.reconfigure(encoding='utf-8')
        sys.stderr.reconfigure(encoding='utf-8')
    except AttributeError:
        pass

# Permitir importaciones de carpetas superiores
sys.path.append(os.path.join(os.path.dirname(__file__), ".."))

from google.genai import types
from db.client import db_query, is_mock_database
from agent.document_parser import parse_document_to_segments, clean_text_content
from agent.retrieval_engine import get_document_category, get_document_owner, ai_client, is_mock_ai

def generate_embedding(text: str) -> list:
    """
    Llama al SDK de Gemini para obtener el embedding del texto,
    o retorna un vector de ceros en modo simulación.
    """
    if is_mock_ai or not ai_client:
        return [0.0] * 1536
    try:
        response = ai_client.models.embed_content(
            model="gemini-embedding-2",
            contents=text,
            config=types.EmbedContentConfig(
                task_type="RETRIEVAL_DOCUMENT",
                output_dimensionality=1536
            )
        )
        return response.embeddings[0].values
    except Exception as e:
        print(f"Error al generar embedding en ingesta: {e}")
        raise e

def sub_chunk_segment(text: str, max_chunk_size: int = 1000, overlap: int = 150) -> list:
    """
    Subdivide texto largo manteniendo solapamiento para mejorar la calidad del RAG.
    """
    if len(text) <= max_chunk_size:
        return [text]
        
    sub_chunks = []
    current_index = 0
    while current_index < len(text):
        end_index = current_index + max_chunk_size
        if end_index < len(text):
            last_space = text.rfind(' ', current_index, end_index)
            if last_space > current_index:
                end_index = last_space
                
        sub_chunk = text[current_index:end_index].strip()
        if sub_chunk:
            sub_chunks.append(sub_chunk)
            
        current_index = end_index - overlap
        if current_index >= len(text) or max_chunk_size <= overlap:
            break
            
    return sub_chunks

def start_ingestion():
    print("Iniciando pipeline de ingesta y extracción RAG en Python...")
    
    kb_dir = os.path.join(os.path.dirname(__file__), "..", "..", "knowledge-base")
    if not os.path.exists(kb_dir):
        print(f"El directorio de base de conocimientos no existe: {kb_dir}")
        sys.exit(1)
        
    files = os.listdir(kb_dir)
    print(f"Archivos detectados en knowledge-base: {', '.join(files)}")
    
    mock_store = []
    
    # Limpiar base de datos si está activa
    if not is_mock_database:
        try:
            print("Limpiando base de datos de fragmentos anteriores...")
            db_query("DELETE FROM document_chunks;")
        except Exception as e:
            print(f"No se pudo limpiar la base de datos: {e}")
            sys.exit(1)
            
    for file in files:
        file_path = os.path.join(kb_dir, file)
        category = get_document_category(file)
        owner = get_document_owner(category)
        
        stat = os.stat(file_path)
        mtime = datetime.datetime.fromtimestamp(stat.st_mtime)
        
        print(f"\nProcesando archivo: '{file}'")
        print(f"- Categoría: {category}")
        print(f"- Responsable: {owner}")
        print(f"- Última Actualización: {mtime.isoformat()}")
        
        try:
            segments = parse_document_to_segments(file_path)
            print(f"Documento dividido originalmente en {len(segments)} segmentos estructurales.")
            
            chunk_counter = 0
            for seg in segments:
                clean_content = clean_text_content(seg["content"])
                if not clean_content:
                    continue
                    
                sub_chunks = sub_chunk_segment(clean_content, 1000, 150)
                for idx, chunk in enumerate(sub_chunks):
                    chunk_counter += 1
                    location = f"{seg['location']} (Parte {idx + 1}/{len(sub_chunks)})" if len(sub_chunks) > 1 else seg["location"]
                    
                    print(f"Indexando fragmento {chunk_counter} ({location})...")
                    
                    # Generar embedding
                    vector = generate_embedding(chunk)
                    
                    if is_mock_database:
                        mock_store.append({
                            "documento_nombre": file,
                            "categoria": category,
                            "contenido": chunk,
                            "ubicacion_exacta": location,
                            "autor_responsable": owner,
                            "ultima_actualizacion": mtime.isoformat(),
                            "embedding": vector
                        })
                    else:
                        vector_string = "[" + ",".join(map(str, vector)) + "]"
                        db_query(
                            """INSERT INTO document_chunks 
                               (documento_nombre, categoria, contenido, ubicacion_exacta, autor_responsable, ultima_actualizacion, embedding)
                               VALUES (%s, %s, %s, %s, %s, %s, %s)""",
                            (file, category, chunk, location, owner, mtime, vector_string)
                        )
                        
            print(f"¡Archivo '{file}' indexado exitosamente con {chunk_counter} fragmentos!")
        except Exception as e:
            print(f"Error crítico procesando el archivo '{file}': {e}")
            
    # Guardar store local si estamos en mock
    if is_mock_database:
        store_path = os.path.join(os.path.dirname(__file__), "..", "..", "logs", "mock_db_vector_store.json")
        os.makedirs(os.path.dirname(store_path), exist_ok=True)
        with open(store_path, "w", encoding="utf-8") as f:
            json.dump(mock_store, f, ensure_ascii=False, indent=2)
        print(f"\n[WARNING] [Base de Datos Mock] Fragmentos locales guardados en cache local: logs/mock_db_vector_store.json")
        
    print('\n=========================================')
    print('¡Pipeline de ingesta finalizado con éxito!')
    print('=========================================')

if __name__ == "__main__":
    start_ingestion()
